File: LyllyPlayer/LyllyPlayer.App/Player/AudioAnalyzer.cs
using NAudio.Dsp;

namespace LyllyPlayer.Player;

public sealed class AudioAnalyzer
{
    /// <summary>Number of log-spaced spectrum buckets (must match visualizer).</summary>
    public const int SpectrumBands = 32;

    // Assumes f32le stereo at 48kHz (matches our ffmpeg args / WaveFormat)
    private const int SampleRate = 48000;
    private const int FftSize = 2048; // power of 2
    /// <summary>Lower edge of the first log band (Hz). Display/analyzer range 10…22050 (CD Nyquist); sub-bin energy still maps into low bands.</summary>
    public const double SpectrumFreqMinHz = 10.0;
    /// <summary>Upper edge of the top band (Hz). <b>22050</b> = 44100/2 (CD Nyquist); aligns with common “full spectrum” labeling.</summary>
    public const double SpectrumFreqMaxHz = 22050.0;

    private readonly object _vuGate = new();
    private readonly object _spectrumGate = new();

    // VU (0..1) — never block on spectrum FFT.
    private float _vuL;
    private float _vuR;

    // Spectrum
    private readonly float[] _ring = new float[FftSize * 4];
    private int _ringWrite;
    private int _ringCount;
    private readonly float[] _bands = new float[SpectrumBands];
    private float _specScale = 1f;
    private long _lastSpectrumComputeTick;

    public void Reset()
    {
        lock (_vuGate)
        {
            _vuL = 0;
            _vuR = 0;
        }

        lock (_spectrumGate)
        {
            Array.Clear(_ring);
            _ringWrite = 0;
            _ringCount = 0;
            Array.Clear(_bands);
            _specScale = 0f;
            _lastSpectrumComputeTick = 0;
        }
    }

    public void ProcessPcmF32LeStereo(byte[] buffer, int offset, int count)
    {
        // 8 bytes per frame (L f32 + R f32); f32le values are already in −1..1 range
        var end = offset + (count - (count % 8));
        if (end <= offset)
            return;

        for (var i = offset; i < end; i += 8)
        {
            var fl = BitConverter.ToSingle(buffer, i);
            var fr = BitConverter.ToSingle(buffer, i + 4);
            var afl = Math.Abs(fl);
            var afr = Math.Abs(fr);

            lock (_vuGate)
            {
                _vuL = SmoothLevel(_vuL, afl);
                _vuR = SmoothLevel(_vuR, afr);
            }

            lock (_spectrumGate)
                WriteRing((fl + fr) * 0.5f);
        }
    }

    public (float vuL, float vuR, float[] bands) GetSnapshot()
    {
        MaybeUpdateSpectrumThrottled();

        float vuL;
        float vuR;
        lock (_vuGate)
        {
            vuL = _vuL;
            vuR = _vuR;
        }

        var copy = new float[SpectrumBands];
        lock (_spectrumGate)
            Array.Copy(_bands, copy, SpectrumBands);
        return (vuL, vuR, copy);
    }

    private void MaybeUpdateSpectrumThrottled()
    {
        var now = Environment.TickCount64;
        if (now - _lastSpectrumComputeTick < 33)
            return;

        lock (_spectrumGate)
        {
            if (_ringCount < FftSize)
                return;
            _lastSpectrumComputeTick = now;
            ComputeSpectrum();
        }
    }

    private static float SmoothLevel(float current, float target)
    {
        const float attack = 0.88f;
        const float release = 0.48f;
        var k = target > current ? attack : release;
        return current + (target - current) * k;
    }

    private void WriteRing(float sample)
    {
        _ring[_ringWrite] = sample;
        _ringWrite = (_ringWrite + 1) % _ring.Length;
        if (_ringCount < _ring.Length) _ringCount++;
    }

    /// <summary>
    /// Lower / upper edge of spectrum band <paramref name="band"/> in Hz (same grid as <see cref="ComputeSpectrum"/>).
    /// Edges follow a <b>geometric progression</b> in Hz — i.e. <b>equal width on a log-frequency axis</b>
    /// (same Δlog f per band), which is how typical log graphic EQs / RTA-style analyzers partition the spectrum — not equal Hz width.
    /// </summary>
    public static void GetSpectrumBandEdgesHz(int band, out double f0Hz, out double f1Hz)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(band, 0, nameof(band));
        ArgumentOutOfRangeException.ThrowIfGreaterThanOrEqual(band, SpectrumBands, nameof(band));

        f0Hz = SpectrumFreqMinHz * Math.Pow(SpectrumFreqMaxHz / SpectrumFreqMinHz, (double)band / SpectrumBands);
        f1Hz = SpectrumFreqMinHz * Math.Pow(SpectrumFreqMaxHz / SpectrumFreqMinHz, (double)(band + 1) / SpectrumBands);
    }

    public static double GetSpectrumBandCenterHz(int band)
    {
        GetSpectrumBandEdgesHz(band, out var f0, out var f1);
        return Math.Sqrt(f0 * f1);
    }

    /// <summary>
    /// Raw FFT energy is a poor match for “how loud it looks” in the subs: hearing is much less sensitive there at
    /// typical listening levels. Down-weight low centers so the graph isn’t dominated by bass you barely perceive.
    /// </summary>
    private static float PerceptualVisibilityWeightHz(float fcHz)
    {
        if (fcHz >= 500f)
            return 1f;
        var lo = MathF.Log10(28f);
        var hi = MathF.Log10(500f);
        var x = MathF.Log10(Math.Clamp(fcHz, 28f, 500f));
        var t = (x - lo) / (hi - lo);
        t = Math.Clamp(t, 0f, 1f);
        const float wLo = 0.34f;
        return wLo + (1f - wLo) * t;
    }

    private void ComputeSpectrum()
    {
        // Pull the latest FftSize samples from ring (mono)
        var temp = new Complex[FftSize];
        var start = (_ringWrite - FftSize + _ring.Length) % _ring.Length;

        for (var i = 0; i < FftSize; i++)
        {
            var s = _ring[(start + i) % _ring.Length];

            // Hann window
            var w = 0.5f - 0.5f * (float)Math.Cos(2 * Math.PI * i / (FftSize - 1));
            temp[i].X = s * w;
            temp[i].Y = 0;
        }

        FastFourierTransform.FFT(true, (int)Math.Log2(FftSize), temp);

        // Convert to magnitudes and bucket into bands (log-ish)
        var mags = new float[FftSize / 2];
        for (var i = 1; i < mags.Length; i++)
        {
            var re = temp[i].X;
            var im = temp[i].Y;
            mags[i] = (float)Math.Sqrt(re * re + im * im);
        }

        // Log-spaced bands 10 Hz…22050 Hz (32 buckets; PCM here is 48 kHz but display matches common CD Nyquist labeling).
        var binDfHz = (float)SampleRate / FftSize;
        var raw = new float[SpectrumBands];
        var frameMax = 1e-6f;
        for (var b = 0; b < SpectrumBands; b++)
        {
            GetSpectrumBandEdgesHz(b, out var f0, out var f1);

            var i0 = (int)Math.Clamp(f0 * FftSize / SampleRate, 1, mags.Length - 1);
            var i1 = (int)Math.Clamp(f1 * FftSize / SampleRate, i0 + 1, mags.Length);

            var n = i1 - i0;
            var sumSq = 0f;
            for (var i = i0; i < i1; i++)
            {
                var m = mags[i];
                sumSq += m * m;
            }

            var rms = (float)Math.Sqrt(sumSq / Math.Max(1, n));

            var bandHz = (float)Math.Max(f1 - f0, 1e-3);
            var narrowBoost = MathF.Sqrt(binDfHz / MathF.Max(bandHz, 0.45f * binDfHz));
            if (narrowBoost < 1f) narrowBoost = 1f;
            if (narrowBoost > 3.4f) narrowBoost = 3.4f;

            var boosted = rms * narrowBoost * 220f;
            var v = (float)Math.Sqrt(boosted);

            raw[b] = v;
            if (v > frameMax) frameMax = v;
        }

        var maxR = frameMax;
        var sorted = new float[SpectrumBands];
        Array.Copy(raw, sorted, SpectrumBands);
        Array.Sort(sorted);
        var p82 = sorted[(int)MathF.Floor(0.82f * (SpectrumBands - 1))];
        var frameRef = MathF.Min(maxR, MathF.Max(p82 * 1.12f, maxR * 0.88f));
        frameRef = MathF.Max(frameRef, 1e-6f);

        const float decay = 0.96f;
        const float minScaleRelativeToPeak = 0.02f;
        const float absScaleFloor = 1e-6f;
        _specScale = Math.Max(frameRef, _specScale * decay);
        var minScale = Math.Max(absScaleFloor, frameRef * minScaleRelativeToPeak);
        if (_specScale < minScale) _specScale = minScale;

        for (var b = 0; b < SpectrumBands; b++)
        {
            var fc = (float)GetSpectrumBandCenterHz(b);
            var w = PerceptualVisibilityWeightHz(fc);
            _bands[b] = Math.Clamp(raw[b] / _specScale * w, 0, 1);
        }
    }
}

File: android/app/src/main/java/com/lyllyplayer/app/visualizer/AudioAnalyzer.kt
package com.lyllyplayer.app.visualizer

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Desktop-aligned spectrum analyzer: 2048 FFT → [SpectrumBands] log-spaced buckets.
 * Feed mono (or mid) PCM floats in −1..1 via [writeSample] / [writeStereoFloat] / [writePcm16].
 */
class AudioAnalyzer {
    private val spectrumGate = Any()
    private val ring = FloatArray(FFT_SIZE * 4)
    private var ringWrite = 0
    private var ringCount = 0
    private val bands = FloatArray(SPECTRUM_BANDS)
    private var specScale = 1f
    private var lastSpectrumComputeTick = 0L
    @Volatile private var sampleRateHz: Int = 48_000

    // Reused FFT scratch (only touched under spectrumGate).
    private val fftRe = FloatArray(FFT_SIZE)
    private val fftIm = FloatArray(FFT_SIZE)
    private val mags = FloatArray(FFT_SIZE / 2)
    private val raw = FloatArray(SPECTRUM_BANDS)
    private val sorted = FloatArray(SPECTRUM_BANDS)

    fun setInputFormat(sampleRateHz: Int) {
        if (sampleRateHz > 0) this.sampleRateHz = sampleRateHz
    }

    fun reset() {
        synchronized(spectrumGate) {
            ring.fill(0f)
            ringWrite = 0
            ringCount = 0
            bands.fill(0f)
            specScale = 0f
            lastSpectrumComputeTick = 0L
        }
    }

    fun writeSample(mono: Float) {
        synchronized(spectrumGate) {
            ring[ringWrite] = mono
            ringWrite = (ringWrite + 1) % ring.size
            if (ringCount < ring.size) ringCount++
        }
    }

    fun writeStereoFloat(left: Float, right: Float) {
        writeSample((left + right) * 0.5f)
    }

    /** Little-endian PCM 16-bit interleaved. */
    fun writePcm16(buffer: ByteArray, offset: Int, length: Int, channelCount: Int) {
        if (channelCount <= 0 || length < 2) return
        val frameBytes = 2 * channelCount
        val end = offset + (length - (length % frameBytes))
        var i = offset
        while (i < end) {
            if (channelCount == 1) {
                val s = pcm16At(buffer, i) / 32768f
                writeSample(s)
            } else {
                val l = pcm16At(buffer, i) / 32768f
                val r = pcm16At(buffer, i + 2) / 32768f
                writeStereoFloat(l, r)
            }
            i += frameBytes
        }
    }

    /** Little-endian float32 interleaved. */
    fun writePcmFloat(buffer: ByteArray, offset: Int, length: Int, channelCount: Int) {
        if (channelCount <= 0 || length < 4) return
        val frameBytes = 4 * channelCount
        val end = offset + (length - (length % frameBytes))
        var i = offset
        while (i < end) {
            if (channelCount == 1) {
                writeSample(floatAt(buffer, i))
            } else {
                writeStereoFloat(floatAt(buffer, i), floatAt(buffer, i + 4))
            }
            i += frameBytes
        }
    }

    /** Copy of current bands (0..1). Throttled FFT ~30 Hz. */
    fun snapshotBands(): FloatArray {
        maybeUpdateSpectrumThrottled()
        val copy = FloatArray(SPECTRUM_BANDS)
        synchronized(spectrumGate) {
            System.arraycopy(bands, 0, copy, 0, SPECTRUM_BANDS)
        }
        return copy
    }

    private fun maybeUpdateSpectrumThrottled() {
        val now = android.os.SystemClock.elapsedRealtime()
        if (now - lastSpectrumComputeTick < 33L) return
        synchronized(spectrumGate) {
            if (ringCount < FFT_SIZE) return
            lastSpectrumComputeTick = now
            computeSpectrum()
        }
    }

    private fun computeSpectrum() {
        val start = (ringWrite - FFT_SIZE + ring.size) % ring.size
        for (i in 0 until FFT_SIZE) {
            val s = ring[(start + i) % ring.size]
            val w = (0.5 - 0.5 * cos(2.0 * PI * i / (FFT_SIZE - 1))).toFloat()
            fftRe[i] = s * w
            fftIm[i] = 0f
        }

        fft(fftRe, fftIm)

        for (i in 1 until mags.size) {
            val re = fftRe[i]
            val im = fftIm[i]
            mags[i] = sqrt(re * re + im * im)
        }

        val rate = sampleRateHz.coerceAtLeast(8_000)
        val binDfHz = rate.toFloat() / FFT_SIZE
        var frameMax = 1e-6f
        for (b in 0 until SPECTRUM_BANDS) {
            val (f0, f1) = spectrumBandEdgesHz(b)
            val i0 = ((f0 * FFT_SIZE / rate).toInt()).coerceIn(1, mags.size - 1)
            val i1 = ((f1 * FFT_SIZE / rate).toInt()).coerceIn(i0 + 1, mags.size)
            var sumSq = 0f
            for (i in i0 until i1) {
                val m = mags[i]
                sumSq += m * m
            }
            val n = (i1 - i0).coerceAtLeast(1)
            val rms = sqrt(sumSq / n)
            val bandHz = max((f1 - f0).toFloat(), 1e-3f)
            var narrowBoost = sqrt(binDfHz / max(bandHz, 0.45f * binDfHz))
            narrowBoost = narrowBoost.coerceIn(1f, 3.4f)
            val v = sqrt(rms * narrowBoost * 220f)
            raw[b] = v
            if (v > frameMax) frameMax = v
        }

        System.arraycopy(raw, 0, sorted, 0, SPECTRUM_BANDS)
        sorted.sort()
        val p82 = sorted[floor(0.82f * (SPECTRUM_BANDS - 1)).toInt()]
        var frameRef = min(frameMax, max(p82 * 1.12f, frameMax * 0.88f))
        frameRef = max(frameRef, 1e-6f)

        specScale = max(frameRef, specScale * 0.96f)
        val minScale = max(1e-6f, frameRef * 0.02f)
        if (specScale < minScale) specScale = minScale

        for (b in 0 until SPECTRUM_BANDS) {
            val fc = spectrumBandCenterHz(b).toFloat()
            val w = perceptualVisibilityWeightHz(fc)
            bands[b] = (raw[b] / specScale * w).coerceIn(0f, 1f)
        }
    }

    companion object {
        const val SPECTRUM_BANDS = 32
        const val SPECTRUM_FREQ_MIN_HZ = 10.0
        const val SPECTRUM_FREQ_MAX_HZ = 22050.0
        private const val FFT_SIZE = 2048

        fun spectrumBandEdgesHz(band: Int): Pair<Double, Double> {
            require(band in 0 until SPECTRUM_BANDS)
            val f0 = SPECTRUM_FREQ_MIN_HZ *
                Math.pow(SPECTRUM_FREQ_MAX_HZ / SPECTRUM_FREQ_MIN_HZ, band.toDouble() / SPECTRUM_BANDS)
            val f1 = SPECTRUM_FREQ_MIN_HZ *
                Math.pow(SPECTRUM_FREQ_MAX_HZ / SPECTRUM_FREQ_MIN_HZ, (band + 1).toDouble() / SPECTRUM_BANDS)
            return f0 to f1
        }

        fun spectrumBandCenterHz(band: Int): Double {
            val (f0, f1) = spectrumBandEdgesHz(band)
            return sqrt(f0 * f1)
        }

        private fun perceptualVisibilityWeightHz(fcHz: Float): Float {
            if (fcHz >= 500f) return 1f
            val lo = log10(28f)
            val hi = log10(500f)
            val x = log10(fcHz.coerceIn(28f, 500f))
            val t = ((x - lo) / (hi - lo)).coerceIn(0f, 1f)
            val wLo = 0.34f
            return wLo + (1f - wLo) * t
        }

        private fun pcm16At(buf: ByteArray, i: Int): Short {
            val lo = buf[i].toInt() and 0xff
            val hi = buf[i + 1].toInt()
            return ((hi shl 8) or lo).toShort()
        }

        private fun floatAt(buf: ByteArray, i: Int): Float {
            val bits = (buf[i].toInt() and 0xff) or
                ((buf[i + 1].toInt() and 0xff) shl 8) or
                ((buf[i + 2].toInt() and 0xff) shl 16) or
                ((buf[i + 3].toInt() and 0xff) shl 24)
            return Float.fromBits(bits)
        }

        /** In-place radix-2 Cooley–Tukey FFT. */
        private fun fft(re: FloatArray, im: FloatArray) {
            val n = re.size
            var j = 0
            for (i in 1 until n) {
                var bit = n shr 1
                while (j and bit != 0) {
                    j = j xor bit
                    bit = bit shr 1
                }
                j = j xor bit
                if (i < j) {
                    val tr = re[i]; re[i] = re[j]; re[j] = tr
                    val ti = im[i]; im[i] = im[j]; im[j] = ti
                }
            }
            var len = 2
            while (len <= n) {
                val ang = -2.0 * PI / len
                val wlenRe = cos(ang).toFloat()
                val wlenIm = sin(ang).toFloat()
                var i0 = 0
                while (i0 < n) {
                    var wRe = 1f
                    var wIm = 0f
                    for (k in 0 until len / 2) {
                        val i = i0 + k
                        val j2 = i + len / 2
                        val tRe = wRe * re[j2] - wIm * im[j2]
                        val tIm = wRe * im[j2] + wIm * re[j2]
                        re[j2] = re[i] - tRe
                        im[j2] = im[i] - tIm
                        re[i] += tRe
                        im[i] += tIm
                        val nextWRe = wRe * wlenRe - wIm * wlenIm
                        wIm = wRe * wlenIm + wIm * wlenRe
                        wRe = nextWRe
                    }
                    i0 += len
                }
                len = len shl 1
            }
        }
    }
}
